kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -902,6 +922,45 @@ private fun ObjCExportMapper.canHaveSameName(first: PropertyDescriptor, second:
     return bridgePropertyType(first) == bridgePropertyType(second)
 }
 
+private fun DeclarationDescriptor.getObjCName(forSwift: Boolean): String {
+    annotations.findAnnotation(KonanFqNames.objCName)?.let { annotation ->
+        fun getName(param: String) = annotation.argumentValue(param)?.value?.cast<String>()?.takeIf(Name::isValidIdentifier)
+        if (forSwift) getName("swiftName")?.let { return it }
+        getName("name")?.let { return it }
+    }
+    return name.asString()
+}
+
+private fun CallableDescriptor.getObjCName(forSwift: Boolean): String =
+        overriddenDescriptors.firstOrNull()?.getObjCName(forSwift) ?: (this as DeclarationDescriptor).getObjCName(forSwift)
```

Comment:
IMO you shouldn't be able to change the name of an override (that isn't possible in Kotlin either).
The same goes for overriding multiple declarations. In Kotlin you can only do that if they have the same signature.
So specifying different `@ObjCName` values for the the base declarations would be an error, right?

Reply:
> IMO you shouldn't be able to change the name of an override (that isn't possible in Kotlin either).

Maybe.
There are circumstances that make the reasoning here harder (also related to https://github.com/JetBrains/kotlin/pull/4815#discussion_r875143243): "base" method in ObjCExport terminology is not the same as "method without `override`". Let's consider a particular example:

```kotlin
internal interface Base1 {
    fun foo()
}

internal interface Base2 {
    fun foo()
}

open class Derived : Base1, Base2 {
    override fun foo() {}
}

class Derived2 : Derived() {
    override fun foo() {}
}
```
`Derived.foo` is a "base" method, `Base*.foo` are not exported.

The question is: how should different combinations of `@ObjCName` on `foo` methods work?
Let's agree for now on the idea that the compiler shouldn't allow `Derived2.foo` to be annotated, or maybe can allow it to have an annotation that matches supers.

Then, I can propose two different designs here:
1. `Derived.foo` should be annotated to change the name.
2. `Base*.foo` should be annotated, and the names should be the same.

Both are not exactly nice: for the first one, when implementing a frontend checker, you wouldn't have access to ObjCExport code to check what is a "base" method; for the second one, frontend checker doesn't need anything from ObjCExport, but would be complicated itself -- it would need to check that two methods match in terms of `ObjCName` annotations on the methods and the parameters.

For `@Throws` annotation we are using the second design, see the checker here: https://github.com/JetBrains/kotlin/blob/a264cbfe7d510edfe65860779468648dfb0f2eb8/native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeThrowsChecker.kt#L33
So I guess for the sake of consistency, it makes sense to prefer the second design, that are not based on ObjCExport "base method" definition.

What do you think?

Reply:
I guess the 2nd option would be the most understandable one. 

In your example `Derived.foo` is considered the "base" method for ObjC. 
However from a Kotlin perspective the `Base*.foo` methods are still considered the base.

Say you add `Derived3`:
```kotlin 
class Derived3 : Base1, Base2 {
    override fun foo() {}
}
```
With the 2nd approach you wouldn't have to think about the `@ObjCName` annotation anymore.

The compiler should definitely produce an error if `Base1.foo` and `Base2.foo` have a different ObjC/Swift name.